### DISCUSSION TITLE
Estimating GNSS Clock Shift

### DISCUSSION THREAD

**Discussion by Galatiy** [groops-devs/groops#75] [Q&A] [answered]
### Description

Hello GROOPS Developers,

I understand that the "[GnssEstimateClockShift](https://groops-devs.github.io/groops/html/GnssEstimateClockShift.html)" estimates the epoch-wise clock shift. The calculation is done with reference to the [inputfileInstrumentRef](https://groops-devs.github.io/groops/html/fileFormat_instrument.html). 

The reference clock should be subtracted from the input clock. With respect to [GNSS PPP](https://groops-devs.github.io/groops/html/cookbook.gnssPpp.html), could you please give me some advice about how I can go about it. 

How can I get a clock bias for each satellite system in the multi-GNSS PPP scenario? For example,  MJD data0, data1, data2 etc? Where data0, ... nth denotes GLO, Galileo, etc. 

Thanks for your understanding and support.

Regards,

Galats.

### Operating systems

- [ ] Linux
- [ ] Windows
- [ ] macOS
- [ ] Other (please specify in the description)

**Comment by zhedumi** [comment]
Hi,

I am a bit unsure of what you want to accomplish so please just correct me if I am wrong in what you want to do. 
the GnssEstimateClockShift programm is there so we can transform observations from different timescales to another time if I remember correctly. It basically just computes the time difference of all satellites from one system to the reference and then computes a mean value over it. 
What you could do is run the program several times for the individual systems. If estimate the clock biases in PPP there should be a related output for the satellites G01. So you could just loop over those and use them in the GnssEstimateClockShift with your reference.


Best regards,

Patrick

**Comment by Galatiy** [comment]
Hi Patrick,

Thanks for your response.

What you have explained is exactly what I was looking for. However, I am not sure how to go about it, more especially incorporating the [GnssEstimateClockShift](https://groops-devs.github.io/groops/html/GnssEstimateClockShift.html) in the script. I, therefore, wonder if you could provide me with a sample script as a starting point.

Thanks for your support.

Regards,

Galats.

**Comment by zhedumi** [comment, marked as answer]
Hi,

I added a small [example.tar.gz](https://github.com/groops-devs/groops/files/9267753/estimateClockShift.tar.gz) on how to gather the bias files and to compute the mean bias of GPS. I also added a second approach instead of 
[GnssEstimateClockShift](https://groops-devs.github.io/groops/html/GnssEstimateClockShift.html) which gives the same result for the mean bias over the same system. Note that the biases are all relativ to the time given within the clock files (first column). In case you want to do another system you just need to change the transmitterList to Galileo or Glonass. 
I hope that somewhat helps in what you want to accomplish.

Best regards,
Patrick

**Comment by Galatiy** [comment]
Hi Patrick,

This will help. If I have some issues, I will let you know. 

Thanks with warm regards,

Galats

**Comment by Waley04** [comment]
Hi Patrick,

In following this discussion, I noted that there are some outliers in the estimated **gpsShift**. I am wondering if you noticed them. 

**===gpsShift.txt===**
[gpsShift.txt](https://github.com/groops-devs/groops/files/9276115/gpsShift.txt)

Here is output example:

**===clockShift_2020-01-01===**
![clockShift_2020-01-01](https://user-images.githubusercontent.com/101952777/183273739-46a8a380-08ee-4178-a835-e5073de275a7.png)

Regards,

Waley.

**Reply by zhedumi** [reply]
Hi waley,

Yea these jumps are quiet normal. They occur in all stations and satellites at the same time due to the zero mean constraint we apply to the observations. We use a zero mean constraint on the GPS clocks which will lead any clock jumps from GPS landing in all satellites and stations as seen in the following figures:
![e01](https://user-images.githubusercontent.com/77068736/183363115-bd704d7b-0adb-4d82-99d2-1dedffec13ca.PNG)
![g01](https://user-images.githubusercontent.com/77068736/183363122-f6d219ab-1dbe-4cfa-97c3-9b6a2f3bfa0d.PNG)
![graz](https://user-images.githubusercontent.com/77068736/183363127-21d4d95c-2845-4fb2-8358-0c699e40d3cc.PNG)
![brux](https://user-images.githubusercontent.com/77068736/183363131-1d7cf849-f690-4572-bf54-fcfd30104f2a.PNG)

They all have the same value. Therefore it is important to use the same products from the same AC when computing PPP.

Best regards,

Patrick

**Comment by Galatiy** [comment]
Dear @zhedumi,

Do you mind sharing the script you used to compute the **clock_transmitter_{LoopTime:%D}.{prn}.dat** in the data directory ([example.tar.gz](https://github.com/groops-devs/groops/files/9267753/estimateClockShift.tar.gz)). 

Thanks in advance. 

Galats.

**Reply by zhedumi** [reply]
Hi,

The clock files I used are form a gnssNetwork estimation I have done. I basically only used [gnssNetwork](https://groops-devs.github.io/groops/html/cookbook.gnssNetwork.html) and set the path for the output in the [clockParametrization](https://groops-devs.github.io/groops/html/gnssParametrizationType#clocks.html).

You can also the final products at IGS and convert them to isntrumentfiles and then do the analysis or have PPP output the estimated receiver clock errors by setting the outputPath within the clockParametrization.

Best regards
